File: week-7/bkryusim/파괴되지_않은_건물.py
def solution(board, skill):
    answer = 0
    behavior = [0, -1, 1]

    dp = [[0 for _ in range(len(board[0]) + 1)] for _ in range(len(board) + 1)]

    for type, r1, c1, r2, c2, degree in skill:
        dp[r1][c1] += behavior[type] * degree
        dp[r1][c2 + 1] -= behavior[type] * degree
        dp[r2 + 1][c1] -= behavior[type] * degree
        dp[r2 + 1][c2 + 1] += behavior[type] * degree

    # 가로방향 부분합
    for r in range(len(board)):
        for c in range(1, len(board[0])):
            dp[r][c] += dp[r][c - 1]

    # 세로방향 부분합
    for c in range(len(board[0])):
        for r in range(1, len(board)):
            dp[r][c] += dp[r - 1][c]

    # 변화율 매트릭스와 기존 board 부호 확인
    for r in range(len(board)):
        for c in range(len(board[0])):
            if board[r][c] + dp[r][c] > 0:
                answer += 1

    return answer

File: week-07/protoseo/파괴되지 않은 건물.java
class Solution {
    int n;
    int m;
    int[][] preSum;
    public int solution(int[][] board, int[][] skill) {
        n = board.length;
        m = board[0].length;
        preSum = new int[n + 1][m + 1];

        for (int[] detail : skill) {
            int type = detail[0];
            int r1 = detail[1];
            int c1 = detail[2];
            int r2 = detail[3];
            int c2 = detail[4];
            int degree = detail[5];

            if (type == 1) {
                preSum[r1][c1] -= degree;
                preSum[r1][c2 + 1] += degree;
                preSum[r2 + 1][c1] += degree;
                preSum[r2 + 1][c2 + 1] -= degree;
            } else if (type == 2) {
                preSum[r1][c1] += degree;
                preSum[r1][c2 + 1] -= degree;
                preSum[r2 + 1][c1] -= degree;
                preSum[r2 + 1][c2 + 1] += degree;
            }
        }

        for (int i = 0; i < m + 1; i++) {
            for (int j = 1; j < n + 1; j++) {
                preSum[j][i] += preSum[j - 1][i];
            }
        }
        for (int i = 0; i < n + 1; i++) {
            for (int j = 1; j < m + 1; j++) {
                preSum[i][j] += preSum[i][j - 1];
            }
        }

        int answer = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] + preSum[i][j] > 0) {
                    answer++;
                }
            }
        }
        return answer;
    }
}
